### ISSUE TITLE
Preprocessing breaks if integration folder is not below package root

### ISSUE THREAD

**Issue by pmelab** [TheBrainFamily/cypress-cucumber-preprocessor#245]
If the integration folder containing `.feature` files is not below the package root where `cypress-cucumber-preprocessor` is installed, preprocessing breaks due to unresolvable dependencies:
```
Error: Cannot find module 'cypress-cucumber-preprocessor/steps' from '[redacted]/cucumber/Cucumber'


This occurred while Cypress was compiling and bundling your test code. This is usually caused by:

- A missing file or dependency
- A syntax error in the file or one of its dependencies
```

Example repository: https://github.com/pmelab/cypress-cucumber-preprocessor-example

I think this could be solved by injecting requirements as globals instead of requiring them from inside the template? I don't know enough about the plugin system and browserify, but I would be happy to have a look if I get some pointers.

**Comment by lgandecki** [comment]
Could you help me understand - in what scenario the .feature would be not below the package root? Is this vuejs specific?

**Comment by pmelab** [comment]
No, this has nothing to do with vue.js. We are building a Drupal extension that provides a central Cypress runtime, so other extensions can provide commands, global step definitions and tests without having to duplicate the cypress configuration. And this brings a couple of "special needs":
1. Since cypress doesn't allow multiple integration folders, we currently scan the Drupal installation for extensions with integration tests and just symlink them into the central cypress integration folder.

2. We would like to control the Cypress version and configuration centrally across projects, which means it should not live in the projects own `package.json` since it might get overridden, cause git conflicts, version conflicts and what not. The project should only provide `.feature` files and step definitions and have no more knowledge about the setup.

3. For local development it is common practice to just symlink a Drupal extension into multiple local versions of Drupal to quickly run tests against multiple versions.

These requirements lead us to a rather "exotic" directory layout:
```
module_a/
  |- integration
module_b/
  |- integration
drupal-8.7/
  |- package.json (without cypress)
  |- .cypress/
    |- package.json (with cypress)
    |- integration/
      |- module_a -> ../../module_a/integration
      |- module_b -> ../../module_b/integration
drupal-8.8/
  ...
drupal-8.6/
  ...
```

So we have situations where the `.feature` file might be outside of the npm package directory containing `cypress-cucumber-processor` or even within *another* package directory, that might or might not have `cypress-cucumber-preprocessor` in its `node_modules`.

This all works just fine with standard cypress tests. But the preprocessing of feature files fail, because the dependency against the `cypress-cucumber-preprocessor` package is resolved against the wrong package directory.

**Comment by lgandecki** [comment]
Ok thanks for the detailed explanatio @pmelab . I'm super busy today, will try to take a closer look tomorrow and at least give you some pointers about getting this to work.

**Comment by lgandecki** [comment]
I didn't look into this as deep as I wanted, but I didn't see a trivial way of making this work.
We do have given/when/then on the global object but they are taken from a template that does require other files from cypress-cucumber-preprocessor anyway, and these file require other things..

Are you able to maybe install the cypress-cucumber-preprocessor globally in your env? That would make it work, I'd assume. but not sure if it's impossible for some other reasons.


It's a really unusual request to have your package accessible outside of the proper node resolution tree. You wouldn't be able to require anything else that comes from a package.json that's below you as well. 

I have a couple ideas, but didn't have enough time to investigate fully. Maybe tomorrow.

**Comment by lgandecki** [comment]
I couldn't help myself..
@pmelab  Try putting this in your devDependencies:
    "cypress-cucumber-preprocessor": "1.17.0-beta",

**Comment by pmelab** [comment]
I works perfectly! Thank you!

The changeset seems surprisingly simple. So the template was always executed in the package context of `cypress-cucumber-preprocessor` anyway?

**Comment by lgandecki** [comment]
@pmelab generation of the template happens in the context of the package - so we can read the dirname then and inject it into the template. When the template is executed, it's in the context of wherever the .feature file is - so if we use absolute path pointing to where the cypress-cucumber package is, it's going to work :) 

Play around it for a couple days and if it continues to work properly I will merge this in.

**Comment by pmelab** [comment]
Hey @lgandecki ! Thanks again. Just wanted to inform you that we`ve been using this for some time in multiple projects now and no more issues turned up.

**Comment by lgandecki** [comment]
Thanks for the confirmation :) You should be able to switch to the latest version now

**Comment by antonio-zenoo** [comment]
Hi, after this change in 1.19.0 I get the following error:

```
The error was:

Error: Cannot find module 'C:Developmentcypress-cucumber-example-master
ode_modulescypress-cucumber-preprocessorlib/resolveStepDefinition' from 'C:\Development\cypress-cucumber-example-master\cypress\integration\socialNetworks'


This occurred while Cypress was compiling and bundling your test code. This is usually caused by:

- A missing file or dependency
- A syntax error in the file or one of its dependencies

Fix the error in your code and re-run your tests.
```

I verified that changing manually the code in cucumberTemplate.js to the previous version, fixes it. What I notice in the error is that the path seems to be missing '/' , as if ${__dirname} would not be getting properly resolved. 

I'm in Windows 10 for more information.

**Comment by mihai-iotic** [comment]
Same issue here. I'm on Windows 10 as well.

**Comment by antonio-zenoo** [comment]
@mihai-iotic for now I solved it by forcing the preprocessor version to 1.18

```
"devDependencies": {
    ...
    "cypress-cucumber-preprocessor": "1.18.0",

```

**Comment by Jscaramal** [comment]
> Hi, after this change in 1.19.0 I get the following error:
> 
> ```
> The error was:
> 
> Error: Cannot find module 'C:Developmentcypress-cucumber-example-master
> ode_modulescypress-cucumber-preprocessorlib/resolveStepDefinition' from 'C:\Development\cypress-cucumber-example-master\cypress\integration\socialNetworks'
> 
> 
> This occurred while Cypress was compiling and bundling your test code. This is usually caused by:
> 
> - A missing file or dependency
> - A syntax error in the file or one of its dependencies
> 
> Fix the error in your code and re-run your tests.
> ```
> 
> I verified that changing manually the code in cucumberTemplate.js to the previous version, fixes it. What I notice in the error is that the path seems to be missing '/' , as if ${__dirname} would not be getting properly resolved.
> 
> I'm in Windows 10 for more information.

This error was fix on 2.0.1 feature

**Comment by vince-ee** [comment]
this is happening again on 4.3.1.


It's trying to lookup the node_modules in directory in relation to where the integrationFolder is.
My integration folder is in another path, as I have a shared framework, and different integration folder (test scenarios ) for different components.

I use `integrationFolder` config to tell Cypress where to find them. 
``` 
   "integrationFolder" : "../../../tests/ui/cypress/integration",
```
